### PR TITLE
Improve Expression with adding type checking functions

### PR DESCRIPTION
**Pull request by cosmicshuai**
1. Remove Array prebuilt-function. You can use createArray funtions.
2. Type Checking prebuilt functions:
isString, isInteger (*), isFloat, isArray, isObject, isDateTime (**), isBoolean

Notes:
\*  In JavaScript, when you call isInteger(1.0), it will true, so isFloat(1.0) will be false.
** isDateTime can will only return true when a ISO TimeStamp string is parsed

### CODE REVIEW

File: libraries/botbuilder-lg/tests/testData/examples/MultiLineExpr.lg [hunk 1]
```diff
@@ -0,0 +1,9 @@
+>Demo How LG can evalute a multiple line defined expression
```
Comment (Danieladu) [inline libraries/botbuilder-lg/tests/testData/examples/MultiLineExpr.lg]: more kinds of test cases especially corner case
if there is no close bracket. 
```
# template
- ${name
- good
```

```
# template
- ${concat('hi', 
   jack)
```

```
# template
[Activity
  key = ${concat('1',
'2'} | hi
```

more...

File: libraries/botbuilder-lg/tests/testData/examples/MultiLineExpr.lg [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+#definition
```
Comment (Danieladu) [inline libraries/botbuilder-lg/tests/testData/examples/MultiLineExpr.lg]: it would be better if we can add one or two test cases in Expression library

File: libraries/adaptive-expressions/src/expressionFunctions.ts [hunk 3]
```diff
@@ -2988,7 +2983,37 @@ export class ExpressionFunctions {
...
                     }),
                 ReturnType.Boolean,
-                ExpressionFunctions.validateIsMatch)
+                ExpressionFunctions.validateIsMatch),
+            
+            // Type Checking Functions
+            new ExpressionEvaluator(ExpressionType.isString, ExpressionFunctions.apply(
+                (args: any[]): boolean =>  typeof args[0] === 'string'),
+            ReturnType.Boolean,
+            ExpressionFunctions.validateUnary),
```
Comment (Danieladu) [inline libraries/adaptive-expressions/src/expressionFunctions.ts]: not just `Unary`, the param's return type should be `string` or `object`.
we can consider to use 
```
validateUnaryString
```

File: libraries/adaptive-expressions/src/expressionFunctions.ts [hunk 4]
```diff
@@ -2988,7 +2983,37 @@ export class ExpressionFunctions {
...
+            new ExpressionEvaluator(ExpressionType.isInteger, ExpressionFunctions.apply(
+                (args: any[]): boolean =>  typeof args[0] === 'number' && args[0] % 1 === 0),
+            ReturnType.Boolean,
+            ExpressionFunctions.validateUnary),
```
Comment (Danieladu) [inline libraries/adaptive-expressions/src/expressionFunctions.ts]: validateUnaryNumber

File: libraries/adaptive-expressions/src/expressionFunctions.ts [hunk 5]
```diff
@@ -2988,7 +2983,37 @@ export class ExpressionFunctions {
...
+            new ExpressionEvaluator(ExpressionType.isFloat, ExpressionFunctions.apply(
+                (args: any[]): boolean =>  typeof args[0] === 'number' && args[0] % 1 !== 0),
```
Comment (Danieladu) [inline libraries/adaptive-expressions/src/expressionFunctions.ts]: what's the difference between Number.isInteger and x % 1 !== 0, we can choose one from them

File: libraries/adaptive-expressions/src/expressionFunctions.ts [hunk 6]
```diff
@@ -2988,7 +2983,37 @@ export class ExpressionFunctions {
...
+            ReturnType.Boolean,
+            ExpressionFunctions.validateUnary),
+            new ExpressionEvaluator(ExpressionType.isArray, ExpressionFunctions.apply(
+                (args: any[]): boolean => Array.isArray(args[0])),
+            ReturnType.Boolean,
+            ExpressionFunctions.validateUnary),
```
Comment (Danieladu) [inline libraries/adaptive-expressions/src/expressionFunctions.ts]: ExpressionFunctions.validateArityAndAnyType(expression, 1, 1, ReturnType.Object);

File: libraries/adaptive-expressions/src/expressionFunctions.ts [hunk 7]
```diff
@@ -2988,7 +2983,37 @@ export class ExpressionFunctions {
...
+            new ExpressionEvaluator(ExpressionType.isObject, ExpressionFunctions.apply(
+                (args: any[]): boolean => typeof args[0] === 'object'),
+            ReturnType.Boolean,
+            ExpressionFunctions.validateUnary),
```
Comment (Danieladu) [inline libraries/adaptive-expressions/src/expressionFunctions.ts]: ExpressionFunctions.validateArityAndAnyType(expression, 1, 1, ReturnType.Object);

File: libraries/adaptive-expressions/src/expressionFunctions.ts [hunk 8]
```diff
@@ -2988,7 +2983,37 @@ export class ExpressionFunctions {
...
+            new ExpressionEvaluator(ExpressionType.isBoolean, ExpressionFunctions.apply(
+                (args: any[]): boolean => typeof args[0] === 'boolean'),
+            ReturnType.Boolean,
+            ExpressionFunctions.validateUnary),
```
Comment (Danieladu) [inline libraries/adaptive-expressions/src/expressionFunctions.ts]: ExpressionFunctions.validateUnaryBoolean

File: libraries/adaptive-expressions/src/expressionFunctions.ts [hunk 9]
```diff
@@ -2988,7 +2983,37 @@ export class ExpressionFunctions {
...
+            new ExpressionEvaluator(ExpressionType.isDateTime, ExpressionFunctions.apply(
+                (args: any[]): boolean => typeof args[0] === 'string' && this.verifyISOTimestamp(args[0]) === undefined),
+            ReturnType.Boolean,
+            ExpressionFunctions.validateUnary)
```
Comment (Danieladu) [inline libraries/adaptive-expressions/src/expressionFunctions.ts]: validateUnaryString

File: libraries/adaptive-expressions/src/expressionFunctions.ts [hunk 10]
```diff
@@ -2988,7 +2983,37 @@ export class ExpressionFunctions {
...
-                ExpressionFunctions.validateIsMatch)
+                ExpressionFunctions.validateIsMatch),
+            
+            // Type Checking Functions
+            new ExpressionEvaluator(ExpressionType.isString, ExpressionFunctions.apply(
+                (args: any[]): boolean =>  typeof args[0] === 'string'),
+            ReturnType.Boolean,
+            ExpressionFunctions.validateUnary),
+            new ExpressionEvaluator(ExpressionType.isInteger, ExpressionFunctions.apply(
+                (args: any[]): boolean =>  typeof args[0] === 'number' && args[0] % 1 === 0),
```
Comment (Danieladu) [inline libraries/adaptive-expressions/src/expressionFunctions.ts]: what's the difference between Number.isInteger and x % 1 !== 0, we can choose one from them

File: libraries/adaptive-expressions/src/expressionFunctions.ts [hunk 11]
```diff
@@ -2988,7 +2983,37 @@ export class ExpressionFunctions {
...
+            ReturnType.Boolean,
+            ExpressionFunctions.validateUnary),
+            new ExpressionEvaluator(ExpressionType.isFloat, ExpressionFunctions.apply(
+                (args: any[]): boolean =>  typeof args[0] === 'number' && args[0] % 1 !== 0),
+            ReturnType.Boolean,
+            ExpressionFunctions.validateUnary),
```
Comment (Danieladu) [inline libraries/adaptive-expressions/src/expressionFunctions.ts]: validateUnaryNumber

File: libraries/adaptive-expressions/src/expressionFunctions.ts (hunk 3 above)
Comment (cosmicshuai) [reply libraries/adaptive-expressions/src/expressionFunctions.ts]: > not just `Unary`, the param's return type should be `string` or `object`.
> we can consider to use
> 
> ```
> validateUnaryString
> ```

My understanding is isString() can take any type argument, and check if the type of args[0] is 'string'. If validateUnaryString used here, once a number is parsed to the function. It will throw an error instead of return false.

Comment (Danieladu) [reply libraries/adaptive-expressions/src/expressionFunctions.ts]: ok make sense

File: libraries/adaptive-expressions/src/expressionFunctions.ts (hunk 10 above)
Comment (Danieladu) [inline libraries/adaptive-expressions/src/expressionFunctions.ts]: use `ExpressionFunctions.isNumber` replace `typeof args[0] === 'number'`

File: libraries/adaptive-expressions/src/expressionFunctions.ts (hunk 5 above)
Comment (Danieladu) [inline libraries/adaptive-expressions/src/expressionFunctions.ts]: use `ExpressionFunctions.isNumber` replace `typeof args[0] === 'number'`
